File: src/pkg10_tf2_cpp/src/listener.cpp
// listener.cpp
// Queries the TF tree: every 0.5 s, look up the transform from "world" to "robot"
// and print the (x, y, yaw) of the robot.

#include <chrono>
#include <memory>

#include "rclcpp/rclcpp.hpp"
#include "tf2_ros/transform_listener.h"
#include "tf2_ros/buffer.h"
#include "tf2/exceptions.h"
#include "tf2/utils.h"
#include "geometry_msgs/msg/transform_stamped.hpp"

using namespace std::chrono_literals;

class TfListener : public rclcpp::Node
{
public:
  TfListener() : Node("tf_listener")
  {
    buffer_   = std::make_unique<tf2_ros::Buffer>(get_clock());
    listener_ = std::make_shared<tf2_ros::TransformListener>(*buffer_);
    timer_    = create_wall_timer(500ms, std::bind(&TfListener::lookup, this));
  }

private:
  void lookup()
  {
    geometry_msgs::msg::TransformStamped t;
    try {
      // lookupTransform(target_frame, source_frame, time)
      // tf2::TimePointZero = "latest available"
      t = buffer_->lookupTransform("world", "robot", tf2::TimePointZero);
    } catch (const tf2::TransformException & ex) {
      RCLCPP_WARN(get_logger(), "Could not transform: %s", ex.what());
      return;
    }

    double yaw = tf2::getYaw(t.transform.rotation);
    RCLCPP_INFO(get_logger(),
      "robot pose in world: x=%.2f y=%.2f yaw=%.2f rad",
      t.transform.translation.x,
      t.transform.translation.y,
      yaw);
  }

  std::unique_ptr<tf2_ros::Buffer> buffer_;
  std::shared_ptr<tf2_ros::TransformListener> listener_;
  rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<TfListener>());
  rclcpp::shutdown();
  return 0;
}

File: src/pkg11_tf2_py/pkg11_tf2_py/listener.py
#!/usr/bin/env python3
"""TF listener in Python: query world -> robot every 0.5 s."""

import math
import rclpy
from rclpy.node import Node
from tf2_ros import TransformException
from tf2_ros.buffer import Buffer
from tf2_ros.transform_listener import TransformListener


def yaw_from_quaternion(q):
    """Extract yaw (rotation about z) from a geometry_msgs Quaternion."""
    siny_cosp = 2.0 * (q.w * q.z + q.x * q.y)
    cosy_cosp = 1.0 - 2.0 * (q.y * q.y + q.z * q.z)
    return math.atan2(siny_cosp, cosy_cosp)


class TfListener(Node):
    def __init__(self):
        super().__init__('tf_listener_py')
        self.buffer = Buffer()
        self.listener = TransformListener(self.buffer, self)
        self.timer = self.create_timer(0.5, self.lookup)

    def lookup(self):
        try:
            t = self.buffer.lookup_transform(
                'world', 'robot', rclpy.time.Time())
        except TransformException as ex:
            self.get_logger().warn(f'Could not transform: {ex}')
            return

        yaw = yaw_from_quaternion(t.transform.rotation)
        self.get_logger().info(
            f'robot pose in world: '
            f'x={t.transform.translation.x:.2f} '
            f'y={t.transform.translation.y:.2f} '
            f'yaw={yaw:.2f} rad')


def main(args=None):
    rclpy.init(args=args)
    node = TfListener()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()
